internal/executor/rpc/artifacts.go

```diff
@@ -34,39 +39,68 @@ func (r *RPC) UploadArtifacts(stream api.CirrusCIService_UploadArtifactsServer)
 				continue
 			}
 
-			artifactPath := filepath.Join(r.artifactsDir, currentArtifactName, obj.Chunk.ArtifactPath)
+			artifactPath := filepath.Join(currentArtifactName, obj.Chunk.ArtifactPath)
 
-			var maybeTruncate int
+			if _, err := artifactWriter.WriteTo(artifactPath, obj.Chunk.Data); err != nil {
+				return err
+			}
+		}
+	}
 
-			if obj.Chunk.ArtifactPath != lastSeenArtifactPath {
-				maybeTruncate = os.O_TRUNC
+	if err := stream.SendAndClose(&api.UploadArtifactsResponse{}); err != nil {
+		r.logger.Warnf("error while closing artifacts stream: %v", err)
+		return err
+	}
 
-				lastSeenArtifactPath = obj.Chunk.ArtifactPath
-			}
+	return nil
+}
 
-			if err := os.MkdirAll(filepath.Dir(artifactPath), 0700); err != nil {
-				return err
-			}
+type ArtifactWriter struct {
+	baseDir     string
+	currentPath string
+	currentFile *os.File
+}
 
-			artifact, err := os.OpenFile(artifactPath, os.O_WRONLY|os.O_APPEND|os.O_CREATE|maybeTruncate, 0600)
-			if err != nil {
-				return err
-			}
+func NewArtifactWriter(baseDir string) *ArtifactWriter {
+	return &ArtifactWriter{
+		baseDir:     baseDir,
+		currentPath: "",
+		currentFile: nil,
+	}
+}
 
-			if _, err := artifact.Write(obj.Chunk.Data); err != nil {
-				return err
-			}
+func (aw *ArtifactWriter) WriteTo(name string, b []byte) (int, error) {
+	path := filepath.Join(aw.baseDir, name)
 
-			if err := artifact.Close(); err != nil {
-				return err
+	// nolint:nestif // doesn't look that complicated
+	if aw.currentFile == nil || aw.currentPath != path {
```

Comment:
Does seem like `aw.currentFile == nil` can be removed. 🤔 

```suggestion
	if aw.currentPath != path {
```